File: app/src/main/java/com/bitchat/android/sync/PacketIdUtil.java
package com.bitchat.android.sync;

import com.bitchat.android.protocol.BitchatPacket;
import com.bitchat.android.util.BinaryEncodingUtils;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

/**
 * Utilitaire pour les ID de paquets déterministes à des fins de synchronisation.
 * Utilise SHA-256 sur un sous-ensemble canonique de champs de paquets.
 */
public final class PacketIdUtil {

    private PacketIdUtil() {}

    public static byte[] computeIdBytes(BitchatPacket packet) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            md.update(packet.getType());
            md.update(packet.getSenderID());

            long ts = packet.getTimestamp();
            for (int i = 7; i >= 0; i--) {
                md.update((byte) ((ts >>> (i * 8)) & 0xFF));
            }

            md.update(packet.getPayload());
            byte[] digest = md.digest();
            return Arrays.copyOf(digest, 16); // ID de 128 bits
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    public static String computeIdHex(BitchatPacket packet) {
        byte[] bytes = computeIdBytes(packet);
        return BinaryEncodingUtils.hexEncodedString(bytes);
    }
}

File: app/src/main/java/com/bitchat/android/sync/PacketIdUtil.kt
package com.bitchat.android.sync

import com.bitchat.android.protocol.BitchatPacket
import com.bitchat.android.util.Hashing
import com.bitchat.android.util.toHexString

/**
 * Deterministic packet ID helper for sync purposes.
 * Uses SHA-256 over a canonical subset of packet fields:
 * [type | senderID | timestamp | payload] to generate a stable ID.
 * Returns a 16-byte (128-bit) truncated hash for compactness.
 */
object PacketIdUtil {
    fun computeIdBytes(packet: BitchatPacket): ByteArray {
        val data = ByteArray(1 + packet.senderID.size + 8 + packet.payload.size)
        var offset = 0
        data[offset++] = packet.type.toByte()
        packet.senderID.copyInto(data, destinationOffset = offset)
        offset += packet.senderID.size

        val ts = packet.timestamp.toLong()
        for (i in 7 downTo 0) {
            data[offset++] = ((ts ushr (i * 8)) and 0xFF).toByte()
        }
        packet.payload.copyInto(data, destinationOffset = offset)

        return Hashing.sha256(data).copyOf(16) // 128-bit ID
    }

    fun computeIdHex(packet: BitchatPacket): String {
        return computeIdBytes(packet).toHexString()
    }
}
